File: internal/gitkit/ts/bin/list-prs.ts
// list-prs.ts — select the current repo's open PRs authored by the viewer that need action.
// Run: node list-prs.ts [--repo <abs path>]  → prints JSON [{pr, headRef, reason}] to stdout.
// --repo (or GIT_SKILL_REPO) anchors the gh calls via repo-root.ts; otherwise ambient cwd.
// Zero npm deps: shells out to `gh`, Node stdlib only. Erasable TS only.

export type PrNode = {
  number: number;
  headRefName: string;
  reviewDecision: string | null;
  author: { login: string } | null;
  reviewThreads: { nodes: { isResolved: boolean }[] };
};

export function countUnresolved(threads: { isResolved: boolean }[]): number {
  return threads.filter((t) => !t.isResolved).length;
}

export function qualifies(reviewDecision: string | null, unresolved: number): boolean {
  return (reviewDecision ?? "").toUpperCase() === "CHANGES_REQUESTED" || unresolved > 0;
}

export function reasonFor(reviewDecision: string | null, unresolved: number): string {
  const parts: string[] = [];
  if ((reviewDecision ?? "").toUpperCase() === "CHANGES_REQUESTED") parts.push("CHANGES_REQUESTED");
  if (unresolved > 0) parts.push(`${unresolved} unresolved thread${unresolved === 1 ? "" : "s"}`);
  return parts.join(" + ");
}

export type Selected = { pr: number; headRef: string; reason: string };

export function selectPrs(nodes: PrNode[], viewerLogin: string): Selected[] {
  const out: Selected[] = [];
  for (const n of nodes) {
    if (n.author?.login !== viewerLogin) continue;
    const unresolved = countUnresolved(n.reviewThreads?.nodes ?? []);
    if (!qualifies(n.reviewDecision, unresolved)) continue;
    out.push({ pr: n.number, headRef: n.headRefName, reason: reasonFor(n.reviewDecision, unresolved) });
  }
  return out;
}

import { execFileSync } from "node:child_process";

import { repoRoot } from "./repo-root.ts";

function sh(cmd: string, args: string[]): string {
  return execFileSync(cmd, args, { encoding: "utf8", maxBuffer: 32 * 1024 * 1024, timeout: 60_000, cwd: repoRoot() });
}

const QUERY = `
query($owner:String!, $repo:String!) {
  viewer { login }
  repository(owner:$owner, name:$repo) {
    pullRequests(states: OPEN, first: 50, orderBy: {field: UPDATED_AT, direction: DESC}) {
      nodes {
        number headRefName reviewDecision
        author { login }
        reviewThreads(first: 100) { nodes { isResolved } }
      }
    }
  }
}`;

async function main(): Promise<void> {
  const repo = JSON.parse(sh("gh", ["repo", "view", "--json", "nameWithOwner"]));
  const [owner, name] = repo.nameWithOwner.split("/");
  const out = sh("gh", ["api", "graphql", "-f", `query=${QUERY}`, "-f", `owner=${owner}`, "-f", `repo=${name}`]);
  const root = JSON.parse(out);
  const viewer: string = root.data.viewer.login;
  const nodes: PrNode[] = root.data.repository.pullRequests.nodes;
  process.stdout.write(JSON.stringify(selectPrs(nodes, viewer), null, 2) + "\n");
}

if (import.meta.main) {
  main().catch((e) => {
    process.stderr.write(`error: ${(e as Error).message}\n`);
    process.exit(1);
  });
}

File: internal/gitkit/listprs.go
package gitkit

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
	"time"
)

// list-prs — the current repo's open PRs authored by the viewer that need
// action (CHANGES_REQUESTED or an unresolved thread): [{pr, headRef, reason}].

type prNode struct {
	Number         int     `json:"number"`
	HeadRefName    string  `json:"headRefName"`
	ReviewDecision *string `json:"reviewDecision"`
	Author         *struct {
		Login string `json:"login"`
	} `json:"author"`
	ReviewThreads *struct {
		Nodes []struct {
			IsResolved bool `json:"isResolved"`
		} `json:"nodes"`
	} `json:"reviewThreads"`
}

func changesRequested(reviewDecision *string) bool {
	return reviewDecision != nil && strings.ToUpper(*reviewDecision) == "CHANGES_REQUESTED"
}

func reasonFor(reviewDecision *string, unresolved int) string {
	parts := []string{}
	if changesRequested(reviewDecision) {
		parts = append(parts, "CHANGES_REQUESTED")
	}
	if unresolved > 0 {
		plural := "s"
		if unresolved == 1 {
			plural = ""
		}
		parts = append(parts, fmt.Sprintf("%d unresolved thread%s", unresolved, plural))
	}
	return strings.Join(parts, " + ")
}

// SelectedPR is one list-prs row, keys in wire order.
type SelectedPR struct {
	PR      int    `json:"pr"`
	HeadRef string `json:"headRef"`
	Reason  string `json:"reason"`
}

func selectPRs(nodes []prNode, viewerLogin string) []SelectedPR {
	out := []SelectedPR{}
	for _, n := range nodes {
		if n.Author == nil || n.Author.Login != viewerLogin {
			continue
		}
		unresolved := 0
		if n.ReviewThreads != nil {
			for _, t := range n.ReviewThreads.Nodes {
				if !t.IsResolved {
					unresolved++
				}
			}
		}
		if !changesRequested(n.ReviewDecision) && unresolved == 0 {
			continue
		}
		out = append(out, SelectedPR{PR: n.Number, HeadRef: n.HeadRefName, Reason: reasonFor(n.ReviewDecision, unresolved)})
	}
	return out
}

const listPRsQuery = `
query($owner:String!, $repo:String!) {
  viewer { login }
  repository(owner:$owner, name:$repo) {
    pullRequests(states: OPEN, first: 50, orderBy: {field: UPDATED_AT, direction: DESC}) {
      nodes {
        number headRefName reviewDecision
        author { login }
        reviewThreads(first: 100) { nodes { isResolved } }
      }
    }
  }
}`

// ghRepo resolves the anchored checkout's owner/name via `gh repo view`.
func ghRepo(opts execOpts) (owner, name string, err error) {
	out, err := execFile(opts, "gh", "repo", "view", "--json", "nameWithOwner")
	if err != nil {
		return "", "", err
	}
	var repo struct {
		NameWithOwner string `json:"nameWithOwner"`
	}
	if err := json.Unmarshal([]byte(out), &repo); err != nil {
		return "", "", err
	}
	owner, name, _ = strings.Cut(repo.NameWithOwner, "/")
	return owner, name, nil
}

func runListPRs(args []string, stdout, stderr io.Writer) int {
	root, err := repoRoot(args)
	if err != nil {
		return fail(stderr, err)
	}
	opts := execOpts{dir: root, echo: stderr, timeout: 60 * time.Second, maxBuffer: 32 << 20}
	owner, name, err := ghRepo(opts)
	if err != nil {
		return fail(stderr, err)
	}
	out, err := execFile(opts, "gh", "api", "graphql", "-f", "query="+listPRsQuery, "-f", "owner="+owner, "-f", "repo="+name)
	if err != nil {
		return fail(stderr, err)
	}
	var resp struct {
		Data *struct {
			Viewer struct {
				Login string `json:"login"`
			} `json:"viewer"`
			Repository *struct {
				PullRequests struct {
					Nodes []prNode `json:"nodes"`
				} `json:"pullRequests"`
			} `json:"repository"`
		} `json:"data"`
	}
	if err := json.Unmarshal([]byte(out), &resp); err != nil {
		return fail(stderr, err)
	}
	if resp.Data == nil || resp.Data.Repository == nil {
		return fail(stderr, errors.New("graphql response carries no repository data"))
	}
	if err := writeJSON(stdout, selectPRs(resp.Data.Repository.PullRequests.Nodes, resp.Data.Viewer.Login)); err != nil {
		return fail(stderr, err)
	}
	return 0
}
